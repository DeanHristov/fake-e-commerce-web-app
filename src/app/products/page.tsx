import { FC } from 'react';
import { Metadata } from 'next';

export interface IProductsProps {}

export const metadata: Metadata = {
  title: 'Products',
};

const ProductsPage: FC<IProductsProps> = ({}) => {
  return <div className="h-[88vh]">Products page</div>;
};

export default ProductsPage;
