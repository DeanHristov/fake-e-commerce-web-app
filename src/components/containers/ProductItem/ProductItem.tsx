'use client';

import Image from 'next/image';
import { FC } from 'react';
import Widget from '@/components/containers/Widget';
import { TCurrency } from '@/types';
import PriceLabel from '@/components/ui/PriceLabel';
import StarRating from '@/components/ui/StarRating';
import StockLabel from '@/components/ui/StockLabel';
import Button from '@/components/ui/Button';

import { HeartIcon, ShoppingCartIcon } from '@heroicons/react/24/solid';
import DiscountLabel from '@/components/ui/DiscountLabel';
import { Utils } from '@/utils/Utils';
import ProductHeading from '@/components/ui/ProductHeading';

export interface IProductItemProps {
  id: number;
  positionIdx: number;
  title: string;
  price: number;
  rating: number;
  category: string;
  thumbnail: string;
  stock: number;
  discountPercentage?: number;
  currency?: TCurrency;
}

const ProductItem: FC<IProductItemProps> = ({
  id,
  positionIdx,
  title,
  price,
  thumbnail,
  rating,
  discountPercentage,
  stock,
  currency = 'EUR',
}) => {
  return (
    <Widget>
      <div className="group mx-auto h-auto max-w-[288px]">
        {Utils.isNotNull(discountPercentage) && (
          <DiscountLabel discount={discountPercentage!} />
        )}
        <div className="w-full rounded-lg border border-gray-300 h-40 flex justify-center overflow-hidden relative">
          <Image
            priority={positionIdx < 10}
            src={thumbnail}
            width={0}
            height={0}
            sizes="100vw"
            alt={title}
            className={`w-full h-auto rounded-lg transition ${
              stock <= 0 && 'group-hover:opacity-80'
            }`}
          />
        </div>
        <ProductHeading id={id} title={title} />
        <PriceLabel
          currency={currency}
          discountPercentage={discountPercentage}
          price={price}
        />
        <StarRating
          activeColor="#FFA133"
          inActiveColor="#E3E3E3"
          starCount={5}
          rating={rating}
        />
        <StockLabel stock={stock} />
        <div className="flex flex-col gap-y-2 pt-2 md:flex-row md:gap-x-2 md:gap-y-0">
          <Button
            variant="secondary"
            leftIcon={<ShoppingCartIcon className={'w-4 h-4 text-dark'} />}
          />
          <Button
            variant="secondary"
            leftIcon={<HeartIcon className={'w-4 h-4 text-dark'} />}
          />
        </div>
      </div>
    </Widget>
  );
};

export default ProductItem;
