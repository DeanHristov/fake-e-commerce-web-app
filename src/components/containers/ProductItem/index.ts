import ProductItem, { IProductItemProps } from './ProductItem';

export type { IProductItemProps };

export default ProductItem;
