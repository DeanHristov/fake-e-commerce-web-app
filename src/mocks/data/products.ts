import { IProduct, PRODUCT_CATEGORIES } from '../../types';

const products: IProduct[] = [
  {
    _id: '29a34b48-8343-4e3d-8c21-3548ba249936',
    name: 'Airpods Pro2 Headphones',
    image: '/images/airpods.jpg',
    description:
      "Apple's new H2 chip is the driving force behind AirPods Pro and their improved sound characteristics. It works in unison with a personalized driver and amplifier for crystal clear heights and deep, rich bass with incredible clarity, so every sound is more pronounced than ever",
    brand: 'Apple',
    category: PRODUCT_CATEGORIES.ELECTRONIC,
    price: 399.99,
    countInStock: 10,
    rating: 4.5,
    currency: 'USD',
    reviews: 5,
  },
];

export default products;
