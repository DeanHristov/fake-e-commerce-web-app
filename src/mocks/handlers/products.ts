import { MockedRequest, rest, RestContext, RestHandler } from 'msw';
import mockedProducts from '../data/products';
import { IProduct } from '@/types';

export let productsHandler: RestHandler<MockedRequest<IProduct>>[] = [
  // Fetching the all products
  rest.get(`${process.env.API_URL}/products`, (req, res, ctx: RestContext) => {
    return res(
      ctx.status(200),
      ctx.json({
        products: mockedProducts,
        total: 100,
        skip: 0,
        limit: 30,
      }),
    );
  }),

  // Fetching a single product
  rest.get(
    `${process.env.API_URL}/products/:productId`,
    (req, res, ctx: RestContext) => {
      const { productId } = req.params;
      const outputItem: IProduct | undefined = mockedProducts.find(
        (item: IProduct) => item.id === Number(productId),
      );

      return res(ctx.status(200), ctx.json(outputItem ?? null));
    },
  ),
];
