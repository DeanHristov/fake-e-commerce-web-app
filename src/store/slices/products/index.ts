export * from './productsSlice';
export * from './selectors';
