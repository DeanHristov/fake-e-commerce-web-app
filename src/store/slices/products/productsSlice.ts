import { IProduct } from '@/types';
import { Utils } from '@/utils/Utils';
import { createSlice, PayloadAction } from '@reduxjs/toolkit';

const reducerName = 'products';
const initialState: Record<string, IProduct> = Utils.tryToLoadFromStorage(
  reducerName,
  {},
);

export const productsSlice = createSlice({
  name: reducerName,
  initialState,
  reducers: {
    initProducts: (
      state: Record<string, IProduct>,
      action: PayloadAction<IProduct[]>,
    ) => {
      for (let productItem of action.payload) {
        state[productItem._id] = productItem;
      }
    },
  },
});

export const { initProducts } = productsSlice.actions;
export default productsSlice;
