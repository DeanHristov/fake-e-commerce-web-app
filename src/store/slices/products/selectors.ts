import { ReduxState } from '@/store';
import { IProduct } from '@/types'; // The function below is called a selector and allows us to select a value from

export const selectProducts = (state: ReduxState): Record<number, IProduct> =>
  state.products;

export const selectProductsById = (
  state: ReduxState,
  productId: number,
): IProduct => state.products[productId];
