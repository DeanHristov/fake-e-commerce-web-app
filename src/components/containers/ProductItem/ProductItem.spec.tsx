import { render } from '@testing-library/react';

import products from '../../../mocks/data/products';
import ProductItem, { IProductItemProps } from './ProductItem';

const mockProduct: IProductItemProps = products.find(
  (item: IProductItemProps) =>
    item._id === '29a34b48-8343-4e3d-8c21-3548ba249936',
)!;

describe('UI/ProductItem <ProductItem {...} />', () => {
  it('Should render the badge with a counter', () => {
    const { container } = render(<ProductItem {...mockProduct} />);

    expect(container).toMatchSnapshot();
  });
});
