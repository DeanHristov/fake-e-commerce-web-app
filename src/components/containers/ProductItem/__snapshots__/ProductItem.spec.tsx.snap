// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UI/ProductItem <ProductItem {...} /> Should render the badge with a counter 1`] = `
<div>
  <div
    class="bg-white border border-gray-300 rounded-lg p-4 shadow-md inline-block relative undefined"
  >
    <div
      class="group mx-auto h-auto max-w-[288px]"
    >
      <div
        class="card-flag-container"
      >
        <span
          class="card-flag"
        >
          -13%
        </span>
      </div>
      <div
        class="w-full rounded-lg border border-gray-300 h-40 flex justify-center overflow-hidden relative"
      >
        <img
          alt="iPhone 9"
          class="w-full h-auto rounded-lg transition false"
          data-nimg="1"
          decoding="async"
          fetchpriority="high"
          height="0"
          sizes="100vw"
          src="/_next/image?url=https%3A%2F%2Fi.dummyjson.com%2Fdata%2Fproducts%2F1%2Fthumbnail.jpg&w=3840&q=75"
          srcset="/_next/image?url=https%3A%2F%2Fi.dummyjson.com%2Fdata%2Fproducts%2F1%2Fthumbnail.jpg&w=640&q=75 640w, /_next/image?url=https%3A%2F%2Fi.dummyjson.com%2Fdata%2Fproducts%2F1%2Fthumbnail.jpg&w=750&q=75 750w, /_next/image?url=https%3A%2F%2Fi.dummyjson.com%2Fdata%2Fproducts%2F1%2Fthumbnail.jpg&w=828&q=75 828w, /_next/image?url=https%3A%2F%2Fi.dummyjson.com%2Fdata%2Fproducts%2F1%2Fthumbnail.jpg&w=1080&q=75 1080w, /_next/image?url=https%3A%2F%2Fi.dummyjson.com%2Fdata%2Fproducts%2F1%2Fthumbnail.jpg&w=1200&q=75 1200w, /_next/image?url=https%3A%2F%2Fi.dummyjson.com%2Fdata%2Fproducts%2F1%2Fthumbnail.jpg&w=1920&q=75 1920w, /_next/image?url=https%3A%2F%2Fi.dummyjson.com%2Fdata%2Fproducts%2F1%2Fthumbnail.jpg&w=2048&q=75 2048w, /_next/image?url=https%3A%2F%2Fi.dummyjson.com%2Fdata%2Fproducts%2F1%2Fthumbnail.jpg&w=3840&q=75 3840w"
          style="color: transparent;"
          width="0"
        />
      </div>
      <h3
        class="product-heading hover:underline hover:cursor-pointer undefined"
      >
        iPhone 9
      </h3>
      <div
        class="space-x-2"
      >
        <span
          class="price-label-old line-through text-sm"
        >
          €620.00
        </span>
        <span
          class="price-label-old"
        >
          /
        </span>
        <span
          class="price-label"
        >
          €549.00
        </span>
      </div>
      <div
        class="flex space-x-2 items-center mt-3"
        data-testid="star-rating"
      >
        <div
          class="flex space-x-1 justify-start items-baseline"
        >
          <span
            class="star-icon"
            data-testid="star-item"
            style="color: rgb(255, 161, 51);"
          >
            ☆
          </span>
          <span
            class="star-icon"
            data-testid="star-item"
            style="color: rgb(255, 161, 51);"
          >
            ☆
          </span>
          <span
            class="star-icon"
            data-testid="star-item"
            style="color: rgb(255, 161, 51);"
          >
            ☆
          </span>
          <span
            class="star-icon"
            data-testid="star-item"
            style="color: rgb(255, 161, 51);"
          >
            ☆
          </span>
          <span
            class="star-icon"
            data-testid="star-item"
            style="color: rgb(255, 161, 51);"
          >
            ☆
          </span>
        </div>
        <span
          class="text-xs font-medium text-zambezi leading-4"
          data-testid="rating"
        >
          4.69 Reviews
        </span>
      </div>
      <div
        class="flex items-center space-x-1 mt-5"
      >
        <div
          class="w-1.5 h-1.5 bg-hover rounded-full"
        />
        <span
          class="text-xs text-black leading-4"
        >
          94+ pcs. in stock
        </span>
      </div>
      <div
        class="flex flex-col gap-y-2 pt-2 md:flex-row md:gap-x-2 md:gap-y-0"
      >
        <button
          class="cursor-pointer bg-white border border-green-600 text-dark font-normal hover:shadow-lg transition transform active:scale-90 duration-200 ease-in w-full px-4 py-1.5 flex justify-center items-center gap-2 undefined"
        >
          <svg
            aria-hidden="true"
            class="w-4 h-4 text-dark"
            fill="currentColor"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M2.25 2.25a.75.75 0 000 1.5h1.386c.17 0 .318.114.362.278l2.558 9.592a3.752 3.752 0 00-2.806 3.63c0 .414.336.75.75.75h15.75a.75.75 0 000-1.5H5.378A2.25 2.25 0 017.5 15h11.218a.75.75 0 00.674-.421 60.358 60.358 0 002.96-7.228.75.75 0 00-.525-.965A60.864 60.864 0 005.68 4.509l-.232-.867A1.875 1.875 0 003.636 2.25H2.25zM3.75 20.25a1.5 1.5 0 113 0 1.5 1.5 0 01-3 0zM16.5 20.25a1.5 1.5 0 113 0 1.5 1.5 0 01-3 0z"
            />
          </svg>
        </button>
        <button
          class="cursor-pointer bg-white border border-green-600 text-dark font-normal hover:shadow-lg transition transform active:scale-90 duration-200 ease-in w-full px-4 py-1.5 flex justify-center items-center gap-2 undefined"
        >
          <svg
            aria-hidden="true"
            class="w-4 h-4 text-dark"
            fill="currentColor"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M11.645 20.91l-.007-.003-.022-.012a15.247 15.247 0 01-.383-.218 25.18 25.18 0 01-4.244-3.17C4.688 15.36 2.25 12.174 2.25 8.25 2.25 5.322 4.714 3 7.688 3A5.5 5.5 0 0112 5.052 5.5 5.5 0 0116.313 3c2.973 0 5.437 2.322 5.437 5.25 0 3.925-2.438 7.111-4.739 9.256a25.175 25.175 0 01-4.244 3.17 15.247 15.247 0 01-.383.219l-.022.012-.007.004-.003.001a.752.752 0 01-.704 0l-.003-.001z"
            />
          </svg>
        </button>
      </div>
    </div>
  </div>
</div>
`;
